<style>
  code {
    font-family: monospace;
    font-size: inherit;
    background: whitesmoke;
    padding: .05rem;
    border-radius: .2rem;
  }
</style>
<p>Aggregates cluster annotated single-cell data into pseudo-bulk data for differential gene expression analysis of the
  respective clusters.</p>
<p>
  <b>Output</b><br />
  Each annotated sample is separated into its clusters and aggregated into a common <code>cluster_pseudobulk.h5ad</code>
  file. Additionally, each annotated sample is separated into data with each cluster removed respectively and aggregated
  into a common <code>cluster_exclude_pseudobulk.h5ad</code> file. Both files contain the following additional
  observation variables:
</p>
<ul>
  <li><code>sample type</code>: the sample type</li>
  <li><code>replicate</code>: the replicate for the specified sample type</li>
  <li><code>clusteringgroup</code>: the grouping that was used for clustering samples</li>
  <li><code>cluster</code>: the cluster that was used for / was excluded from pseudo-bluk generation</li>
  <li><code>cellcount</code>: the number of cells that were used to create the pseudo-bulk sample</li>
</ul>

<p><b>References</b></p>
<ol>
  <li>
    Heumos, L., Schaar, A.C., Lance, C. et al. Best practices for single-cell
    analysis across modalities. Nat Rev Genet (2023).
    <a href="https://doi.org/10.1038/s41576-023-00586-w">
      https://doi.org/10.1038/s41576-023-00586-w
    </a>
  </li>
</ol>