<p>Calculates differentially expressed genes between different sample types.</p>
<p><b>References</b></p>
<ol>
  <li>
    Gagnon J, Pi L, Ryals M, Wan Q, Hu W, Ouyang Z, Zhang B, Li K. Recommendations of scRNA-seq Differential Gene
    Expression Analysis Based on Comprehensive Benchmarking. Life. 2022; 12(6):850.
    <a href="https://doi.org/10.3390/life12060850">
      https://doi.org/10.3390/life12060850
    </a>
  </li>
  <li>
    Squair, J.W., Gautier, M., Kathe, C. et al. Confronting false discoveries in single-cell differential expression.
    Nat Commun 12, 5692 (2021).
    <a href="https://doi.org/10.1038/s41467-021-25960-2">
      https://doi.org/10.1038/s41467-021-25960-2
    </a>
  </li>
  <li>
    Heumos, L., Schaar, A.C., Lance, C. et al. Best practices for single-cell
    analysis across modalities. Nat Rev Genet (2023).
    <a href="https://doi.org/10.1038/s41576-023-00586-w">
      https://doi.org/10.1038/s41576-023-00586-w
    </a>
  </li>
</ol>