<style>
  code {
    font-family: monospace;
    font-size: inherit;
    background: whitesmoke;
    padding: .05rem;
    border-radius: .2rem;
  }
</style>
<p>Calculates differentially expressed genes of specific clusters and of clusteres between sample types.</p>
<p>
  <b>Output</b><br />
  For each clustering group and each cluster a respective subfolder is generated at
  <code>cluster_comparison/{clustering_group}/{cluster}</code>. Similarly, for each sample type / condition comparison
  specified in the <code>sample_comparison.csv</code> and each cluster a respective subfolder is generated at
  <code>sample_comparison/{sample_type_1}__vs__{sample_type_2}/{cluster}</code>. Each subfolder contains the following
  files:
<ul>
  <li><code>differential_gene_expression.csv</code>: a CSV containing differentially expressed genes including log2 fold
    change, p value and FDR</li>
  <li><code>info.csv</code>: a CSV file containing information related to the differential gene expression analysis</li>
  <li><code>mds_plot_plot.svg</code>: a multi-dimensional scaling plot of the pseudobulk data (similar samples are
    expected to be located in close proximity)</li>
  <li><code>bcv_plot.svg</code>: a biological coefficient of variation plot that shows gene expression variability</li>
  <li><code>smear_plot.svg</code>: a smear plot that shows change in expression against abundance</li>
</ul>
</p>

<p><b>References</b></p>
<ol>
  <li>
    Gagnon J, Pi L, Ryals M, Wan Q, Hu W, Ouyang Z, Zhang B, Li K. Recommendations of scRNA-seq Differential Gene
    Expression Analysis Based on Comprehensive Benchmarking. Life. 2022; 12(6):850.
    <a href="https://doi.org/10.3390/life12060850">
      https://doi.org/10.3390/life12060850
    </a>
  </li>
  <li>
    Squair, J.W., Gautier, M., Kathe, C. et al. Confronting false discoveries in single-cell differential expression.
    Nat Commun 12, 5692 (2021).
    <a href="https://doi.org/10.1038/s41467-021-25960-2">
      https://doi.org/10.1038/s41467-021-25960-2
    </a>
  </li>
  <li>
    Heumos, L., Schaar, A.C., Lance, C. et al. Best practices for single-cell
    analysis across modalities. Nat Rev Genet (2023).
    <a href="https://doi.org/10.1038/s41576-023-00586-w">
      https://doi.org/10.1038/s41576-023-00586-w
    </a>
  </li>
</ol>