<style>
  code {
    font-family: monospace;
    font-size: inherit;
    background: whitesmoke;
    padding: 0.05rem;
    border-radius: 0.2rem;
  }
</style>
<p>Clusters separated samples at a default resolution.</p>
<p>
  <b>Output</b><br />
  For each clustering group the following files are generated:
</p>
<ul>
  <li>
    <code>clustered.h5ad</code>: the concatenated sample data including
    clustering information
  </li>
  <li>
    <code>umap.svg</code>: a UMAP visualisation of the clustering, sample types
    and replicates
  </li>
  <li>
    <code>marker_genes.svg</code>: a dotplot showing marker genes for the
    respective clusteres
  </li>
</ul>
<p><b>References</b></p>
<ol>
  <li>
    Wolf, F. A., Angerer, P., & Theis, F. J. (2018). SCANPY: large-scale
    single-cell gene expression data analysis. Genome biology, 19(1), 15.
    <a href="https://doi.org/10.1186/s13059-017-1382-0">
      https://doi.org/10.1186/s13059-017-1382-0
    </a>
  </li>
  <li>
    Virshup, I., Rybakov, S., Theis, F. J., Angerer, P., & Wolf, F. A. (2021). anndata: Annotated data (p.
    2021.12.16.473007). bioRxiv.
    <a href="https://doi.org/10.1101/2021.12.16.473007">
      https://doi.org/10.1101/2021.12.16.473007
    </a>
  </li>
  <li>
    Heumos, L., Schaar, A.C., Lance, C. et al. Best practices for single-cell
    analysis across modalities. Nat Rev Genet (2023).
    <a href="https://doi.org/10.1038/s41576-023-00586-w">
      https://doi.org/10.1038/s41576-023-00586-w
    </a>
  </li>
</ol>