<p>Splits the alignment into nucleosome specific chuncks.</p>
<p><b>References</b></p>
<ol>
    <li>Buenrostro, J. D., Giresi, P. G., Zaba, L. C., Chang, H. Y., & Greenleaf, W. J. (2013). Transposition of native
        chromatin for fast and sensitive epigenomic profiling of open chromatin, DNA-binding proteins and nucleosome
        position. Nature methods, 10(12), 1213â€“1218. <a
            href="https://doi.org/10.1038/nmeth.2688">https://doi.org/10.1038/nmeth.2688</a></li>
    <li><a href="https://github.com/pysam-developers/pysam">pysam</a>: htslib interface for python</li>
    <li>HTSlib: C library for reading/writing high-throughput sequencing data. James K Bonfield, John Marshall, Petr
        Danecek, Heng Li, Valeriu Ohan, Andrew Whitwham, Thomas Keane, Robert M Davies. GigaScience, Volume 10, Issue 2,
        February 2021, giab007, <a
            href="https://doi.org/10.1093/gigascience/giab007">https://doi.org/10.1093/gigascience/giab007</a></li>
    <li>Twelve years of SAMtools and BCFtools. Petr Danecek, James K Bonfield, Jennifer Liddle, John Marshall, Valeriu
        Ohan, Martin O Pollard, Andrew Whitwham, Thomas Keane, Shane A McCarthy, Robert M Davies, Heng Li. GigaScience,
        Volume 10, Issue 2, February 2021, giab008, <a
            href="https://doi.org/10.1093/gigascience/giab008">https://doi.org/10.1093/gigascience/giab008</a></li>
</ol>