<style>
  code {
    font-family: monospace;
    font-size: inherit;
    background: whitesmoke;
    padding: 0.05rem;
    border-radius: 0.2rem;
  }
</style>
<p>Estimates the effective / mappable genome size.</p>
<p>
  <b>Output</b>
</p>
<p>
  A report file <code>unique_kmer_report.txt</code> containing unique kmer information is created per sample directory.
  This represents an estimate for the effective / mappable genome size.
</p>
<p>
  <b>References</b>
</p>
<ol>
  <li>
    Crusoe, M. R., Alameldin, H. F., Awad, S., Boucher, E., Caldwell, A., Cartwright, R., Charbonneau, A.,
    Constantinides, B., Edvenson, G., Fay, S., Fenton, J., Fenzl, T., Fish, J., Garcia-Gutierrez, L., Garland, P.,
    Gluck, J., González, I., Guermond, S., Guo, J., Gupta, A., … Brown, C. T. (2015). The khmer software package:
    enabling efficient nucleotide sequence analysis. F1000Research, 4, 900.
    <a href="https://doi.org/10.12688/f1000research.6924.1">https://doi.org/10.12688/f1000research.6924.1</a>
  </li>
  <li>
    Döring, A., Weese, D., Rausch, T., & Reinert, K. (2008). SeqAn an efficient, generic C++ library for sequence
    analysis. BMC bioinformatics, 9, 11.
    <a href="https://doi.org/10.1186/1471-2105-9-11">https://doi.org/10.1186/1471-2105-9-11</a>
  </li>
  <li>
    Irber, L. C., & Brown, C. T. (2016). Efficient cardinality estimation for k-mers in large DNA sequencing data sets.
    bioRxiv, 056846.
    <a href="https://doi.org/10.1101/056846 ">https://doi.org/10.1101/056846 </a>
  </li>
  </li>
</ol>