<style>
  code {
    font-family: monospace;
    font-size: inherit;
    background: whitesmoke;
    padding: 0.05rem;
    border-radius: 0.2rem;
  }
</style>
<p>Performs pathway analysis for each sample comparison.</p>
<p>
  <b>Output</b><br />
  For each sample comparison and used pathway databasea
  <code>pathway_analysis_{pathway database}_table.csv</code>s is produced. It contains the following columns:
</p>
<ul>
  <li><code>pathway</code>: the ID of the respective pathway</li>
  <li><code>score</code>: the enrichment score of the pathway</li>
  <li><code>pvalue</code>: the p-value for the enrichment</li>
</ul>
<br />
<p>
  Additionally, an extended table
  <code>pathway_analysis_{pathway database}_genes.csv</code> is produced that contains
  differential peak accesibility information for every signifanctly altered
  pathway. It contains the following columns:
</p>
<ul>
  <li><code>source</code>: the ID of the respective pathway</li>
  <li><code>target</code>: the symbol of the associated gene</li>
  <li><code>weight</code>: the influence strength of the associated gene on the pathway</li>
  <li><code>log2FoldChange</code>: the log-2-fold change in accessibility</li>
  <li><code>pvalue</code>: the p-value for the enrichment</li>
  <li><code>padj</code>: the adjusted p-value for the enrichment</li>
  <li><code>PeakID</code>: the ID of the respcetive peak</li>
</ul>
<br />
<p>
  Finally, a barplot <code>pathway_analysis_{pathway database}_barplot.svg</code> is created
  showing significant (p-value &le; 0.05) pathway alterations.
</p>
<p><b>References</b></p>
<ol>
  <li>
    Badia-i-Mompel P., Vélez Santiago J., Braunger J., Geiss C., Dimitrov D.,
    Müller-Dott S., Taus P., Dugourd A., Holland C.H., Ramirez Flores R.O. and
    Saez-Rodriguez J. 2022. decoupleR: ensemble of computational methods to
    infer biological activities from omics data. Bioinformatics Advances.
    <a href="https://doi.org/10.1093/bioadv/vbac016">
      https://doi.org/10.1093/bioadv/vbac016
    </a>
  </li>
  <li>
    Schubert, M., Klinger, B., Klünemann, M. et al. Perturbation-response genes reveal signaling footprints in cancer
    gene expression. Nat Commun 9, 20 (2018).
    <a href="https://doi.org/10.1038/s41467-017-02391-6">
      https://doi.org/10.1038/s41467-017-02391-6
    </a>
  </li>
  <li>
    Hunter, J. D. (2007). Matplotlib: A 2D graphics environment. Computing in
    Science & Engineering, 9(3), 90–95.
    <a href="https://doi.org/10.1109/MCSE.2007.55">https://doi.org/10.1109/MCSE.2007.55</a>
  </li>
  <li>
    Waskom, M. L. (2021). seaborn: Statistical data visualization. Journal of
    Open Source Software, 6(60), 3021.
    <a href="https://doi.org/10.21105/joss.03021">https://doi.org/10.21105/joss.03021</a>
  </li>
  <li>
    McKinney, W. (2010). Data Structures for Statistical Computing in Python. In
    S. van der Walt & J. Millman (Hrsg.), Proceedings of the 9th Python in
    Science Conference (S. 56–61).
    <a href="https://doi.org/10.25080/Majora-92bf1922-00a">https://doi.org/10.25080/Majora-92bf1922-00a</a>
  </li>
  <li>
    The pandas development team. (2020). pandas-dev/pandas: Pandas (latest)
    [Software]. Zenodo.
    <a href="https://doi.org/10.5281/zenodo.3509134">https://doi.org/10.5281/zenodo.3509134</a>
  </li>
</ol>