<style>
  code {
    font-family: monospace;
    font-size: inherit;
    background: whitesmoke;
    padding: .05rem;
    border-radius: .2rem;
  }
</style>
<p>Performs an initial quality control of sequenced reads and removes low quality cells.</p>
<p>
  <b>Output</b><br />
  For each sample the preprocessed <code>filtered_feature_bc_matrix.h5ad</code> file is generated.
  Metrics are summed up in the <code>metrics.csv</code> file. Additionally the following files are generated per sample:
<ul>
  <li><code>ddqc_box_plot_genes.svg</code>: Boxplot of the number of genes per QC cluster. The conventionally used
    thershold of 200 genes is marked with a red line.</li>
  <li><code>ddqc_box_plot_mitochondrial_reads.svg</code>: Boxplot of the percentage of mitochondiral reads per QC
    cluster. The conventionally used thershold of 10 % is marked with a red line.</li>
  <li><code>ddqc_line_plot_absolute.svg</code>: Line plot of the number of cells per QC cluster that would be filtered
    out when setting the MAD threashold to the respective x-axis value. The threshold is marked with a dotted line.</li>
  <li><code>ddqc_line_plot_relative.svg</code>: Line plot of the percentage of cells per QC cluster that would be
    filtered out when setting the MAD threashold to the respective x-axis value. The threshold is marked with a dotted
    line.</li>
  <li><code>ddqc_table.csv</code>: Table containg the QC results. For further information see the <a
      href="https://github.com/ayshwaryas/ddqc/tree/master">ddqc documentation</a>.</li>
</ul>
</p>
<p>
  <b>References</b>
</p>
<ol>
  <li>Subramanian, A., Alperovich, M., Yang, Y., & Li, B. (2022). Biology-inspired data-driven quality control for
    scientific discovery in single-cell transcriptomics. Genome Biology, 23, 267. <a
      href="https://doi.org/10.1186/s13059-022-02820-w">https://doi.org/10.1186/s13059-022-02820-w</a>
  </li>
  <li>Li, B., Gould, J., Yang, Y., Sarkizova, S., Tabaka, M., Ashenberg, O., Rosen, Y., Slyper, M., Kowalczyk, M. S.,
    Villani, A. C., Tickle, T., Hacohen, N., Rozenblatt-Rosen, O., & Regev, A. (2020). Cumulus provides cloud-based data
    analysis for large-scale single-cell and single-nucleus RNA-seq. Nature methods, 17(8), 793–798. <a
      href="https://doi.org/10.1038/s41592-020-0905-x">https://doi.org/10.1038/s41592-020-0905-x</a>
  </li>

  <li>Heumos, L., Schaar, A.C., Lance, C. et al. Best practices for single-cell analysis across modalities. Nat Rev
    Genet (2023). <a href="https://doi.org/10.1038/s41576-023-00586-w">https://doi.org/10.1038/s41576-023-00586-w</a>
  </li>
  <li>
    Wolf, F. A., Angerer, P., & Theis, F. J. (2018). SCANPY: large-scale
    single-cell gene expression data analysis. Genome biology, 19(1), 15.
    <a href="https://doi.org/10.1186/s13059-017-1382-0">
      https://doi.org/10.1186/s13059-017-1382-0
    </a>
  </li>
  <li>
    Virshup, I., Rybakov, S., Theis, F. J., Angerer, P., & Wolf, F. A. (2021).
    anndata: Annotated data (p. 2021.12.16.473007). bioRxiv.
    <a href="https://doi.org/10.1101/2021.12.16.473007">
      https://doi.org/10.1101/2021.12.16.473007
    </a>
  </li>
  <li>
    Hunter, J. D. (2007). Matplotlib: A 2D graphics environment. Computing in
    Science & Engineering, 9(3), 90–95.
    <a href="https://doi.org/10.1109/MCSE.2007.55">https://doi.org/10.1109/MCSE.2007.55</a>
  </li>
  <li>
    Waskom, M. L. (2021). seaborn: Statistical data visualization. Journal of
    Open Source Software, 6(60), 3021.
    <a href="https://doi.org/10.21105/joss.03021">https://doi.org/10.21105/joss.03021</a>
  </li>
  <li>
    McKinney, W. (2010). Data Structures for Statistical Computing in Python. In
    S. van der Walt & J. Millman (Hrsg.), Proceedings of the 9th Python in
    Science Conference (S. 56–61).
    <a href="https://doi.org/10.25080/Majora-92bf1922-00a">https://doi.org/10.25080/Majora-92bf1922-00a</a>
  </li>
  <li>
    The pandas development team. (2020). pandas-dev/pandas: Pandas (latest)
    [Software]. Zenodo.
    <a href="https://doi.org/10.5281/zenodo.3509134">https://doi.org/10.5281/zenodo.3509134</a>
  </li>
  <li>
    Virtanen, P., Gommers, R., Oliphant, T. E., Haberland, M., Reddy, T.,
    Cournapeau, D., Burovski, E., Peterson, P., Weckesser, W., Bright, J., van
    der Walt, S. J., Brett, M., Wilson, J., Millman, K. J., Mayorov, N., Nelson,
    A. R. J., Jones, E., Kern, R., Larson, E., … SciPy 1.0 Contributors. (2020).
    SciPy 1.0: Fundamental Algorithms for Scientific Computing in Python. Nature
    Methods, 17, 261–272.
    <a href="https://doi.org/10.1038/s41592-019-0686-2">https://doi.org/10.1038/s41592-019-0686-2</a>
  </li>
</ol>