<p>Normalises merged data using different methods for downstream analysis.</p>
<p>
    <b>Output</b><br />
    A single <code>normalised.h5ad</code> file is generated containing various normalisations of the data.
</p>
<p><b>References</b></p>
<ol>
    <li>
        RA Dorfman. A note on the !d-method for finding variance formulae. Biometric Bulletin, 1938.
    </li>
    <li>
        Germain, PL., Sonrel, A. & Robinson, M.D. pipeComp, a general framework for the evaluation of computational
        pipelines, reveals performant single cell RNA-seq preprocessing tools. Genome Biol 21, 227 (2020).
        <a href="https://doi.org/10.1186/s13059-020-02136-7">
            https://doi.org/10.1186/s13059-020-02136-7
        </a>
    </li>
    <li>
        Lun ATL, McCarthy DJ, Marioni JC (2016). A step-by-step workflow for low-level analysis of single-cell RNA-seq
        data with Bioconductor. F1000Res., 5, 2122.
        <a href="doi:10.12688/f1000research.9501.2">
            doi:10.12688/f1000research.9501.2
        </a>
    </li>
    <li>
        Heumos, L., Schaar, A.C., Lance, C. et al. Best practices for single-cell analysis across modalities. Nat Rev
        Genet (2023).
        <a href="https://doi.org/10.1038/s41576-023-00586-w">
            https://doi.org/10.1038/s41576-023-00586-w
        </a>
    </li>
    <li>
        Booeshaghi, A. S., Hallgrímsdóttir, I. B., Gálvez-Merchán, Á., & Pachter, L. (2022). Depth normalization for
        single-cell genomics count data (p. 2022.05.06.490859). bioRxiv.
        <a href="https://doi.org/10.1101/2022.05.06.490859">
            https://doi.org/10.1101/2022.05.06.490859
        </a>
    </li>
    <li>
        Hunter, J. D. (2007). Matplotlib: A 2D graphics environment. Computing in
        Science & Engineering, 9(3), 90–95.
        <a href="https://doi.org/10.1109/MCSE.2007.55">https://doi.org/10.1109/MCSE.2007.55</a>
    </li>
    <li>
        Waskom, M. L. (2021). seaborn: Statistical data visualization. Journal of
        Open Source Software, 6(60), 3021.
        <a href="https://doi.org/10.21105/joss.03021">https://doi.org/10.21105/joss.03021</a>
    </li>
    <li>
        McKinney, W. (2010). Data Structures for Statistical Computing in Python. In
        S. van der Walt & J. Millman (Hrsg.), Proceedings of the 9th Python in
        Science Conference (S. 56–61).
        <a href="https://doi.org/10.25080/Majora-92bf1922-00a">https://doi.org/10.25080/Majora-92bf1922-00a</a>
    </li>
    <li>
        The pandas development team. (2020). pandas-dev/pandas: Pandas (latest)
        [Software]. Zenodo.
        <a href="https://doi.org/10.5281/zenodo.3509134">https://doi.org/10.5281/zenodo.3509134</a>
    </li>
    <li>
        Virtanen, P., Gommers, R., Oliphant, T. E., Haberland, M., Reddy, T.,
        Cournapeau, D., Burovski, E., Peterson, P., Weckesser, W., Bright, J., van
        der Walt, S. J., Brett, M., Wilson, J., Millman, K. J., Mayorov, N., Nelson,
        A. R. J., Jones, E., Kern, R., Larson, E., … SciPy 1.0 Contributors. (2020).
        SciPy 1.0: Fundamental Algorithms for Scientific Computing in Python. Nature
        Methods, 17, 261–272.
        <a href="https://doi.org/10.1038/s41592-019-0686-2">https://doi.org/10.1038/s41592-019-0686-2</a>
    </li>
</ol>